xla/service/gpu/runtime/nccl_collective_permute_thunk.h

```diff
@@ -46,26 +46,34 @@ class NcclCollectivePermuteStartThunk : public NcclCollectiveThunk {
   NcclCollectivePermuteStartThunk(ThunkInfo thunk_info, NcclApi* nccl_api,
                                   const HloCollectivePermuteInstruction* instr,
                                   int64_t replica_count,
-                                  int64_t partition_count,
-                                  const Buffer& buffer);
+                                  int64_t partition_count, const Buffer& buffer,
+                                  bool p2p_memcpy_enabled);
+  absl::Status Initialize(const InitializeParams& params) override;
+
+  absl::Status Cleanup(const CleanupParams& params) override;
 
   static const char* GetHloOpName() { return "collective-permute-start"; }
 
  protected:
   const NcclCollectiveConfig& config() const override { return config_.config; }
   absl::Status RunNcclCollective(const ExecuteParams& params,
                                  se::Stream& stream,
-                                 NcclApi::NcclCommHandle comm) override;
+                                 NcclApi::NcclCommHandle comm,
+                                 bool is_local) override;
 
  private:
   const NcclP2PConfig config_;
   const Buffer buffer_;
+  std::unordered_map<int64_t, uint64_t> send_value_map_;
```

Comment:
unordered map doesn't guarantee pointer stability, you can easily get asan errors in deregister, you need absl::node_hash_map here

Reply:
IIUC, references and pointers to elements stored in unordered map dont get invalidated except in the case of removal, only iterators are invalidated in cases of rehashing, so i think the stability should be guaranteed. However I looked into the implementation of node_hash_map, looks like it's more optimal in terms of storage so i changed it to node_hash_map anyway.